R/tidy_single_line_address.R

```diff
@@ -2,31 +2,70 @@
-                                                                                                              "&", " AND "),        # replace the ampersand character with the string "and"
-                                                                                                              "( ){2,}", " "),      # replace any multiple spaces with a single space
-                                                                                                              " - ", "-")           # remove any spaces around a hyphen
-                                                                                                              )                     # Trim whitespace
-    )
+  # Check if object is a lazy table
+  if(inherits(df, c("tbl_dbi", "tbl_lazy"))){
+
+    # Process as a lazy frame
+    df %>%
+      dplyr::mutate(
+        {{ col }} := trimws(REPLACE(REGEXP_REPLACE(REPLACE(REGEXP_REPLACE(REGEXP_REPLACE(REGEXP_REPLACE(toupper({{ col }}),             # Uppercase
+                                                                                                                "[,.();:#'']", " "),      # replace special characters with a single space
+                                                                                                                "(\\d)(\\D)", "\\1 \\2"), # add a space between any digit followed by a non-digit (e.g. 1A becomes 1 A)
+                                                                                                                "(\\D)(\\d)", "\\1 \\2"), # add a space between any non-digit followed by a digit (e.g. A1 becomes A 1)
+                                                                                                                "&", " AND "),            # replace the ampersand character with the string "and"
+                                                                                                                "( ){2,}", " "),          # replace any multiple spaces with a single space
+                                                                                                                " - ", "-")               # remove any spaces around a hyphen
+        ),
+
+        # Only remove spaces around hyphen if surrounded by numbers
+        {{ col }} := case_when(
+          REGEXP_INSTR({{ col }}, "[0-9] - [0-9]") > 0L ~ REPLACE(" - ", "-", {{ col }}),
+          TRUE ~ {{ col }}
+        )
+      )
+  }else{
+
+    #Process as a data frame
+    df %>%
+      mutate(
+        # Address cleaning
+        {{ col }} := toupper({{ col }}),
+        {{ col }} := gsub(" & ", " AND ", {{ col }}),
+        {{ col }} := gsub("(\\D)(\\d)", "\\1 \\2", {{ col }}),
+        {{ col }} := gsub("(\\d)(\\D)", "\\1 \\2", {{ col }}),
+        {{ col }} := gsub("[,.();:#''\"]", " ", {{ col }}),
+        {{ col }} := stringr::str_squish({{ col }}),
+        {{ col }} := ifelse(
```

Comment:
Note for later: looked into using `gsub("(?<=\\d)\\s-\\s(?=\\d)", "-", {{ col }}, perl = TRUE)`. This appeared to work fine, but unlike the db version could not test on a huge dataset. It may be useful at some point...